core/src/main/scala/org/scalafmt/util/TreeOps.scala

```diff
@@ -460,4 +460,24 @@ object TreeOps {
   // procedure syntax has decltpe: Some("")
   def isProcedureSyntax(defn: Defn.Def): Boolean =
     defn.decltpe.exists(_.tokens.isEmpty)
+
+  def isCaseClass(tree: Tree): Boolean = {
+    tree.is[Defn.Class] && tree.children.exists(_.is[Mod.Case]) ||
+      tree.parent.exists(_.is[Defn.Class]) && tree.parent.exists(_.children.exists(_.is[Mod.Case]))
+  }
+
+  def isCaseObject(tree: Tree): Boolean = {
+    tree.is[Defn.Object] && tree.children.exists(_.is[Mod.Case]) ||
+      tree.parent.exists(_.is[Defn.Object]) && tree.parent.exists(_.children.exists(_.is[Mod.Case]))
```

Comment:
Here is how I would implement this

```scala
def isCaseObject(tree: Tree): Boolean = tree match {
  case defn: Defn.Object => defn.mods.exists(_.is[Mod.Case])
  case _ => false
}
```

The current implementation would return true for a `Mod` or `Ctor.Primary`, which is just false.

If you have a `mod: Mod`, you can do `mod.parent.exists(isCaseObject)`.

Think of `Tree` as `Object` or `Any`, we should avoid working with it where possible.